src/main/java/soot/baf/toolkits/base/OpStackCalculator.java

```diff
@@ -0,0 +1,863 @@
+//      return true;
+//    }
+//    if (t1 instanceof DoubleType && t2 instanceof DoubleType) {
+//      return true;
+//    }
+//    if (t1 instanceof FloatType && t2 instanceof FloatType) {
+//      return true;
+//    }
+//    return false;
+//  }
+
+  private static RefType isHandlerUnit(Chain<Trap> traps, Unit h) {
+    for (Trap t : traps) {
+      if (t.getHandlerUnit() == h) {
+        return t.getException().getType();
+      }
+    }
+    return null;
+  }
+
+  private static void printStack(StackEffectSwitch sw, PatchingChain<Unit> units, Map<Unit, Stack<Type>> stacks) {
+    try {
+      sw.shouldThrow = false;
+
+      Map<Unit, Integer> indexes = new HashMap<Unit, Integer>();
+      {
+        int count = 0;
+        for (Iterator<Unit> it = units.snapshotIterator(); it.hasNext();) {
+          indexes.put(it.next(), count++);
+        }
+      }
+      for (Iterator<Unit> it = units.snapshotIterator(); it.hasNext();) {
+        Unit unit = it.next();
+        String s = "";
+        if (unit instanceof TargetArgInst) {
+          Unit t = ((TargetArgInst) unit).getTarget();
+          s = indexes.get(t).toString();
+        } else if (unit instanceof TableSwitchInst) {
+          TableSwitchInst tswi = (TableSwitchInst) unit;
+          s += "\r\tdefault: " + tswi.getDefaultTarget() + "  " + indexes.get(tswi.getDefaultTarget());
+          int index = 0;
+          for (int x = tswi.getLowIndex(), e = tswi.getHighIndex(); x <= e; x++) {
+            s += "\r\t " + x + ": " + tswi.getTarget(index) + "  " + indexes.get(tswi.getTarget(index++));
+          }
+        }
+        try {
+          s = indexes.get(unit) + " " + unit + "  " + s + "   [";
+        } catch (Exception e) {
+          logger.debug("Error in OpStackCalculator trying to find index of unit");
+        }
+        Stack<Type> stack = stacks.get(unit);
+        if (stack != null) {
+          unit.apply(sw);
+          stack = updateStack(sw, stack);
+          if (stack == null) {
+            printUnits(units, " StackTypeHeightCalc failed");
+            return;
+          }
+          for (int i = 0, e = stack.size(); i < e; i++) {
+            s += printType(stack.get(i));
+          }
+        } else {
+          s += "***missing***";
+        }
+        System.out.println(s + "]");
+      }
+    } finally {
+      sw.shouldThrow = true;
+    }
+  }
+
+  private static String printType(Type t) {
+    if (t instanceof IntegerType) {
+      return "I";
+    } else if (t instanceof FloatType) {
+      return "F";
+    } else if (t instanceof DoubleType) {
+      return "D";
+    } else if (t instanceof LongType) {
+      return "J";
+    } else if (t instanceof RefLikeType) {
+      return "L" + t.toString();
+    } else {
+      return "U(" + t.getClass().toString() + ")";
+    }
+  }
+
+  public static void printUnits(PatchingChain<Unit> u, String msg) {
+    System.out.println("\r\r***********  " + msg);
```

Comment:
looks like this is just for your debugging purpose?
Are you aware of this class?
https://github.com/soot-oss/soot/blob/51acd8db18bb90b91f469c282ac49a62b9438c6b/src/main/java/soot/Printer.java

Reply:
Like I said earlier, OpStackCalculator is based on soot.jbco.bafTransformations.StackTypeHeightCalculator but with some modifications and decoupling from the jbco package so I really didn't put any effort into changing the debugging aspects.